missing required field, ModifyTargetGroupAttributesInput.Attributes.

Hello, I am trying the last build that allows configurable target groups (#275) without success.

Here is the bug report:

**Summary**
The alb-ingress-controller can't correctly configure the target group attributes and, because of that, It can't continue to create target groups.

**Steps to reproduce**
Configure an ALB Ingress resource with following annotations at least :
```
alb.ingress.kubernetes.io/target-group-attributes: "stickiness.enabled=true"
```
**Version**
Build: 109f7c7

**Expected behaviour**
It's expected to get all target groups successfully configured with the sticky sessions enabled.

**Additional Information**
I tried also adding all attributes to the annotation without success.

Log output:
```
E0621 11:11:51.360825       1 session.go:44] [ALB-INGRESS] [session] [ERROR]: Failed request: elasticloadbalancing/ModifyTargetGroupAttributes, Payload: {  TargetGroupArn: "arn:aws:elasticloadbalancing:eu-west-1:194174110123:targetgroup/0004199999-30902-HTTP-075a0e7/b640b04c31ff3167"}, Error: InvalidParameter: 1 validation error(s) found.
E0621 11:11:51.360978       1 session.go:44] [ALB-INGRESS] [session] [ERROR]: - missing required field, ModifyTargetGroupAttributesInput.Attributes.
E0621 11:11:51.361012       1 session.go:44] [ALB-INGRESS] [session] [ERROR]: 
I0621 11:11:51.361079       1 targetgroup.go:227] [ALB-INGRESS] [default/alb-ingress] [INFO]: Failed TargetGroup creation. Unable to add target group attributes: InvalidParameter: 1 validation error(s) found.
I0621 11:11:51.361227       1 targetgroup.go:227] [ALB-INGRESS] [default/alb-ingress] [INFO]: - missing required field, ModifyTargetGroupAttributesInput.Attributes.
I0621 11:11:51.361708       1 targetgroup.go:227] [ALB-INGRESS] [default/alb-ingress] [INFO]: .
E0621 11:11:51.361796       1 albingress.go:332] [ALB-INGRESS] [default/alb-ingress] [ERROR]: Failed to reconcile state on this ingress
E0621 11:11:51.362155       1 albingress.go:334] [ALB-INGRESS] [default/alb-ingress] [ERROR]:  - InvalidParameter: 1 validation error(s) found.
E0621 11:11:51.362272       1 albingress.go:334] [ALB-INGRESS] [default/alb-ingress] [ERROR]: - missing required field, ModifyTargetGroupAttributesInput.Attributes.
E0621 11:11:51.362332       1 albingress.go:334] [ALB-INGRESS] [default/alb-ingress] [ERROR]: 
I0621 11:11:51.362535       1 controller.go:486] ingress backend successfully reloaded...
I0621 11:11:51.363431       1 event.go:218] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"default", Name:"alb-ingress", UID:"e31915cd-7543-11e8-aae1-0a30a642a36c", APIVersion:"extensions", ResourceVersion:"5325", FieldPath:""}): type: 'Warning' reason: 'ERROR' Error adding attributes to target group 0004199999-30902-HTTP-075a0e7: InvalidParameter: 1 validation error(s) found.
- missing required field, ModifyTargetGroupAttributesInput.Attributes.
I0621 11:12:12.450814       1 status.go:337] updating Ingress default/alb-ingress status to [{ 0004199999-default-albingr-d740-1057553044.eu-west-1.elb.amazonaws.com}]
I0621 11:12:12.455279       1 event.go:218] Event(v1.ObjectReference{Kind:"Ingress", Namespace:"default", Name:"alb-ingress", UID:"e31915cd-7543-11e8-aae1-0a30a642a36c", APIVersion:"extensions", ResourceVersion:"5363", FieldPath:""}): type: 'Normal' reason: 'UPDATE' Ingress default/alb-ingress
I0621 11:12:51.949536       1 alb-controller.go:145] [ALB-INGRESS] [controller] [INFO]: Ingress update being attempted. (Forced from no event seen in 60 seconds).
I0621 11:12:51.950329       1 session.go:35] [ALB-INGRESS] [session] [INFO]: Request: elasticloadbalancing/DescribeTargetHealth, Payload: {  TargetGroupArn: "arn:aws:elasticloadbalancing:eu-west-1:194174110123:targetgroup/0004199999-30902-HTTP-075a0e7/b640b04c31ff3167",  Targets: [{      Id: "i-09248e01b586040c6",      Port: 30902    },{      Id: "i-0926d07bc3dc73f1d",      Port: 30902    },{      Id: "i-0d459efae49ed92ae",      Port: 30902    }]}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x162c4c8]

goroutine 73 [running]:
github.com/coreos/alb-ingress-controller/pkg/alb/targetgroups.NewDesiredTargetGroups(0xc42086fbe8, 0xc420168010, 0x2, 0x2, 0x3, 0xc420610e60)
        /home/travis/gopath/src/github.com/coreos/alb-ingress-controller/pkg/alb/targetgroups/targetgroups.go:172 +0x798
github.com/coreos/alb-ingress-controller/pkg/albingress.NewALBIngressFromIngress(0xc420af9f20, 0x27212e0, 0xc4202a8900, 0x0)
        /home/travis/gopath/src/github.com/coreos/alb-ingress-controller/pkg/albingress/albingress.go:166 +0x982
github.com/coreos/alb-ingress-controller/pkg/albingresses.NewALBIngressesFromIngresses(0xc42086fee8, 0x27212e0, 0xc4202a8900, 0x1, 0xc420665130, 0xc420710870)
        /home/travis/gopath/src/github.com/coreos/alb-ingress-controller/pkg/albingresses/albingresses.go:74 +0x29f
github.com/coreos/alb-ingress-controller/pkg/controller.(*albController).update(0xc420268280)
        /home/travis/gopath/src/github.com/coreos/alb-ingress-controller/pkg/controller/alb-controller.go:196 +0x2d9
github.com/coreos/alb-ingress-controller/pkg/controller.startPolling(0xc420268280)
        /home/travis/gopath/src/github.com/coreos/alb-ingress-controller/pkg/controller/alb-controller.go:146 +0x10d
created by github.com/coreos/alb-ingress-controller/pkg/controller.(*albController).Configure
        /home/travis/gopath/src/github.com/coreos/alb-ingress-controller/pkg/controller/alb-controller.go:129 +0x2ef
```

--- a/pkg/alb/listeners/listeners.go
+++ b/pkg/alb/listeners/listeners.go
@@ -48,15 +48,17 @@ func (ls Listeners) Reconcile(rOpts *ReconcileOptions) (Listeners, error) {
 			return nil, err
 		}
 
-		rsOpts := &rules.ReconcileOptions{
-			Eventf:       rOpts.Eventf,
-			ListenerArn:  l.Current.ListenerArn,
-			TargetGroups: rOpts.TargetGroups,
-		}
-		if rs, err := l.Rules.Reconcile(rsOpts); err != nil {
-			return nil, err
-		} else {
-			l.Rules = rs
+		if l.Current != nil {
+			rsOpts := &rules.ReconcileOptions{
+				Eventf:       rOpts.Eventf,
+				ListenerArn:  l.Current.ListenerArn,
+				TargetGroups: rOpts.TargetGroups,
+			}
+			if rs, err := l.Rules.Reconcile(rsOpts); err != nil {
+				return nil, err
+			} else {
+				l.Rules = rs
+			}
 		}
 		if !l.Deleted {
 			output = append(output, l)

--- a/pkg/alb/loadbalancer/loadbalancer.go
+++ b/pkg/alb/loadbalancer/loadbalancer.go
@@ -116,7 +116,6 @@ func NewDesiredLoadBalancer(o *NewDesiredLoadBalancerOptions) *LoadBalancer {
 		newLoadBalancer.DesiredInboundCidrs = o.Annotations.InboundCidrs
 	}
 
-	// TODO: What is this for??
 	if o.ExistingLoadBalancer != nil {
 		// we had an existing LoadBalancer in ingress, so just copy the desired state over
 		o.ExistingLoadBalancer.Desired = newLoadBalancer.Desired

--- a/pkg/alb/rule/rule.go
+++ b/pkg/alb/rule/rule.go
@@ -138,7 +138,7 @@ func (r *Rule) Reconcile(rOpts *ReconcileOptions) error {
 			log.Prettify(r.Current.Conditions))
 
 	default:
-		r.logger.Debugf("No listener modification required.")
+		r.logger.Debugf("No rule modification required.")
 	}
 
 	return nil

--- a/pkg/alb/targetgroup/targetgroup.go
+++ b/pkg/alb/targetgroup/targetgroup.go
@@ -31,8 +31,8 @@ type TargetGroup struct {
 	ID                string
 	SvcName           string
 	Tags              Tags
-	CurrentAttributes []*elbv2.TargetGroupAttribute
-	DesiredAttributes []*elbv2.TargetGroupAttribute
+	CurrentAttributes albelbv2.TargetGroupAttributes
+	DesiredAttributes albelbv2.TargetGroupAttributes
 	Current           *elbv2.TargetGroup
 	Desired           *elbv2.TargetGroup
 	Targets           Targets
@@ -42,14 +42,23 @@ type TargetGroup struct {
 
 type NewDesiredTargetGroupOptions struct {
 	Annotations    *annotations.Annotations
-	Attributes     []*elbv2.TargetGroupAttribute
 	Tags           util.Tags
 	ALBNamePrefix  string
 	LoadBalancerID string
 	Port           int64
 	Logger         *log.Logger
 	SvcName        string
+	Targets        util.AWSStringSlice
 }
+
+type tgChange uint
+
+const (
+	paramsModified tgChange = 1 << iota
+	targetsModified
+	tagsModified
+	attributesModified
+)
 
 // NewDesiredTargetGroup returns a new targetgroup.TargetGroup based on the parameters provided.
 func NewDesiredTargetGroup(o *NewDesiredTargetGroupOptions) *TargetGroup {
@@ -85,6 +94,9 @@ func NewDesiredTargetGroup(o *NewDesiredTargetGroupOptions) *TargetGroup {
 		Tags: Tags{
 			Desired: newTagList,
 		},
+		Targets: Targets{
+			Desired: o.Targets,
+		},
 		Desired: &elbv2.TargetGroup{
 			HealthCheckPath:            o.Annotations.HealthcheckPath,
 			HealthCheckIntervalSeconds: o.Annotations.HealthcheckIntervalSeconds,
@@ -100,7 +112,7 @@ func NewDesiredTargetGroup(o *NewDesiredTargetGroupOptions) *TargetGroup {
 			UnhealthyThresholdCount: o.Annotations.UnhealthyThresholdCount,
 			// VpcId:
 		},
-		DesiredAttributes: o.Attributes,
+		DesiredAttributes: o.Annotations.TargetGroupAttributes,
 	}
 }
 
@@ -158,20 +170,20 @@ func (tg *TargetGroup) Reconcile(rOpts *ReconcileOptions) error {
 		tg.logger.Infof("Succeeded TargetGroup creation. ARN: %s | Name: %s.",
 			*tg.Current.TargetGroupArn,
 			*tg.Current.TargetGroupName)
-
-		// Current and Desired exist and need for modification should be evaluated.
-	case tg.needsModification():
-		tg.logger.Infof("Start TargetGroup modification.")
-		if err := tg.modify(rOpts); err != nil {
-			return err
-		}
-		rOpts.Eventf(api.EventTypeNormal, "CREATE", "%s target group modified", tg.ID)
-		tg.logger.Infof("Succeeded TargetGroup modification. ARN: %s | Name: %s.",
-			*tg.Current.TargetGroupArn,
-			*tg.Current.TargetGroupName)
-
 	default:
-		tg.logger.Debugf("No TargetGroup modification required.")
+		// Current and Desired exist and need for modification should be evaluated.
+		if m := tg.needsModification(); m != 0 {
+			tg.logger.Infof("Start TargetGroup modification.")
+			if err := tg.modify(m, rOpts); err != nil {
+				return err
+			}
+			rOpts.Eventf(api.EventTypeNormal, "CREATE", "%s target group modified", tg.ID)
+			tg.logger.Infof("Succeeded TargetGroup modification. ARN: %s | Name: %s.",
+				*tg.Current.TargetGroupArn,
+				*tg.Current.TargetGroupName)
+		} else {
+			tg.logger.Debugf("No TargetGroup modification required.")
+		}
 	}
 
 	return nil
@@ -217,24 +229,27 @@ func (tg *TargetGroup) create(rOpts *ReconcileOptions) error {
 		tg.logger.Infof("Failed TargetGroup creation. Unable to register targets:  %s.", err.Error())
 		return err
 	}
+
 	// Add TargetGroup attributes
 	attributes := &elbv2.ModifyTargetGroupAttributesInput{
 		Attributes:     tg.DesiredAttributes,
 		TargetGroupArn: tg.Current.TargetGroupArn,
 	}
+
 	if _, err := albelbv2.ELBV2svc.ModifyTargetGroupAttributes(attributes); err != nil {
 		rOpts.Eventf(api.EventTypeWarning, "ERROR", "Error adding attributes to target group %s: %s", tg.ID, err.Error())
 		tg.logger.Infof("Failed TargetGroup creation. Unable to add target group attributes: %s.", err.Error())
 		return err
 	}
+	tg.CurrentAttributes = tg.DesiredAttributes
+
 	return nil
 }
 
 // Modifies the attributes of an existing TargetGroup.
 // ALBIngress is only passed along for logging
-func (tg *TargetGroup) modify(rOpts *ReconcileOptions) error {
-	// check/change attributes
-	if tg.needsModification() {
+func (tg *TargetGroup) modify(m tgChange, rOpts *ReconcileOptions) error {
+	if m&paramsModified != 0 {
 		in := &elbv2.ModifyTargetGroupInput{
 			HealthCheckIntervalSeconds: tg.Desired.HealthCheckIntervalSeconds,
 			HealthCheckPath:            tg.Desired.HealthCheckPath,
@@ -259,7 +274,7 @@ func (tg *TargetGroup) modify(rOpts *ReconcileOptions) error {
 	}
 
 	// check/change tags
-	if *tg.Tags.Current.Hash() != *tg.Tags.Desired.Hash() {
+	if m&tagsModified != 0 {
 		if err := albelbv2.ELBV2svc.UpdateTags(tg.Current.TargetGroupArn, tg.Tags.Current, tg.Tags.Desired); err != nil {
 			rOpts.Eventf(api.EventTypeWarning, "ERROR", "Error changing tags on target group %s: %s", tg.ID, err.Error())
 			tg.logger.Errorf("Failed TargetGroup modification. Unable to modify tags. ARN: %s | Error: %s.",
@@ -269,30 +284,29 @@ func (tg *TargetGroup) modify(rOpts *ReconcileOptions) error {
 		tg.Tags.Current = tg.Tags.Desired
 	}
 
-	additions := util.Difference(tg.Targets.Desired, tg.Targets.Current)
-	removals := util.Difference(tg.Targets.Current, tg.Targets.Desired)
+	if m&targetsModified != 0 {
+		additions := util.Difference(tg.Targets.Desired, tg.Targets.Current)
+		removals := util.Difference(tg.Targets.Current, tg.Targets.Desired)
 
-	// check/change targets
-	if len(additions) > 0 {
-		if err := tg.registerTargets(additions, rOpts); err != nil {
-			rOpts.Eventf(api.EventTypeWarning, "ERROR", "Error adding targets to target group %s: %s", tg.ID, err.Error())
-			tg.logger.Infof("Failed TargetGroup modification. Unable to add targets: %s.", err.Error())
-			return err
+		// check/change targets
+		if len(additions) > 0 {
+			if err := tg.registerTargets(additions, rOpts); err != nil {
+				rOpts.Eventf(api.EventTypeWarning, "ERROR", "Error adding targets to target group %s: %s", tg.ID, err.Error())
+				tg.logger.Infof("Failed TargetGroup modification. Unable to add targets: %s.", err.Error())
+				return err
+			}
 		}
-	}
-	if len(removals) > 0 {
-		if err := tg.deregisterTargets(removals, rOpts); err != nil {
-			rOpts.Eventf(api.EventTypeWarning, "ERROR", "Error removing targets from target group %s: %s", tg.ID, err.Error())
-			tg.logger.Infof("Failed TargetGroup modification. Unable to remove targets: %s.", err.Error())
-			return err
+		if len(removals) > 0 {
+			if err := tg.deregisterTargets(removals, rOpts); err != nil {
+				rOpts.Eventf(api.EventTypeWarning, "ERROR", "Error removing targets from target group %s: %s", tg.ID, err.Error())
+				tg.logger.Infof("Failed TargetGroup modification. Unable to remove targets: %s.", err.Error())
+				return err
+			}
 		}
+		tg.Targets.Current = tg.Targets.Desired
 	}
 
-	// check/change attributes
-	currentAttributes := albelbv2.TargetGroupAttributes(tg.CurrentAttributes).Sorted()
-	desiredAttributes := albelbv2.TargetGroupAttributes(tg.DesiredAttributes).Sorted()
-
-	if !reflect.DeepEqual(currentAttributes, desiredAttributes) {
+	if m&attributesModified != 0 {
 		attributes := &elbv2.ModifyTargetGroupAttributesInput{
 			Attributes:     tg.DesiredAttributes,
 			TargetGroupArn: tg.Current.TargetGroupArn,
@@ -302,6 +316,7 @@ func (tg *TargetGroup) modify(rOpts *ReconcileOptions) error {
 			tg.logger.Infof("Failed TargetGroup modification. Unable to change attributes: %s.", err.Error())
 			return err
 		}
+		tg.CurrentAttributes = tg.DesiredAttributes
 	}
 
 	return nil
@@ -316,48 +331,76 @@ func DeleteTG(tg *TargetGroup) error {
 	return nil
 }
 
-func (tg *TargetGroup) needsModification() bool {
+func (tg *TargetGroup) needsModification() tgChange {
+	var changes tgChange
+
 	ctg := tg.Current
 	dtg := tg.Desired
 
-	switch {
 	// No target group set currently exists; modification required.
-	case ctg == nil:
+	if ctg == nil {
 		tg.logger.Debugf("Current Target Group is undefined")
-		return true
-	case !util.DeepEqual(ctg.HealthCheckIntervalSeconds, dtg.HealthCheckIntervalSeconds):
-		tg.logger.Debugf("HealthCheckIntervalSeconds needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckIntervalSeconds), log.Prettify(dtg.HealthCheckIntervalSeconds))
-		return true
-	case !util.DeepEqual(ctg.HealthCheckPath, dtg.HealthCheckPath):
-		tg.logger.Debugf("HealthCheckPath needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckPath), log.Prettify(dtg.HealthCheckPath))
-		return true
-	case !util.DeepEqual(ctg.HealthCheckPort, dtg.HealthCheckPort):
-		tg.logger.Debugf("HealthCheckPort needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckPort), log.Prettify(dtg.HealthCheckPort))
-		return true
-	case !util.DeepEqual(ctg.HealthCheckProtocol, dtg.HealthCheckProtocol):
-		tg.logger.Debugf("HealthCheckProtocol needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckProtocol), log.Prettify(dtg.HealthCheckProtocol))
-		return true
-	case !util.DeepEqual(ctg.HealthCheckTimeoutSeconds, dtg.HealthCheckTimeoutSeconds):
-		tg.logger.Debugf("HealthCheckTimeoutSeconds needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckTimeoutSeconds), log.Prettify(dtg.HealthCheckTimeoutSeconds))
-		return true
-	case !util.DeepEqual(ctg.HealthyThresholdCount, dtg.HealthyThresholdCount):
-		tg.logger.Debugf("HealthyThresholdCount needs to be changed (%v != %v)", log.Prettify(ctg.HealthyThresholdCount), log.Prettify(dtg.HealthyThresholdCount))
-		return true
-	case !util.DeepEqual(ctg.Matcher, dtg.Matcher):
-		tg.logger.Debugf("Matcher needs to be changed (%v != %v)", log.Prettify(ctg.Matcher), log.Prettify(ctg.Matcher))
-		return true
-	case !util.DeepEqual(ctg.UnhealthyThresholdCount, dtg.UnhealthyThresholdCount):
-		tg.logger.Debugf("UnhealthyThresholdCount needs to be changed (%v != %v)", log.Prettify(ctg.UnhealthyThresholdCount), log.Prettify(dtg.UnhealthyThresholdCount))
-		return true
-	case *tg.Targets.Current.Hash() != *tg.Targets.Desired.Hash():
-		tg.logger.Debugf("Targets need to be changed.")
-		return true
+		return changes
 	}
+
+	if !util.DeepEqual(ctg.HealthCheckIntervalSeconds, dtg.HealthCheckIntervalSeconds) {
+		tg.logger.Debugf("HealthCheckIntervalSeconds needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckIntervalSeconds), log.Prettify(dtg.HealthCheckIntervalSeconds))
+		changes |= paramsModified
+	}
+
+	if !util.DeepEqual(ctg.HealthCheckPath, dtg.HealthCheckPath) {
+		tg.logger.Debugf("HealthCheckPath needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckPath), log.Prettify(dtg.HealthCheckPath))
+		changes |= paramsModified
+	}
+
+	if !util.DeepEqual(ctg.HealthCheckPort, dtg.HealthCheckPort) {
+		tg.logger.Debugf("HealthCheckPort needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckPort), log.Prettify(dtg.HealthCheckPort))
+		changes |= paramsModified
+	}
+
+	if !util.DeepEqual(ctg.HealthCheckProtocol, dtg.HealthCheckProtocol) {
+		tg.logger.Debugf("HealthCheckProtocol needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckProtocol), log.Prettify(dtg.HealthCheckProtocol))
+		changes |= paramsModified
+	}
+
+	if !util.DeepEqual(ctg.HealthCheckTimeoutSeconds, dtg.HealthCheckTimeoutSeconds) {
+		tg.logger.Debugf("HealthCheckTimeoutSeconds needs to be changed (%v != %v)", log.Prettify(ctg.HealthCheckTimeoutSeconds), log.Prettify(dtg.HealthCheckTimeoutSeconds))
+		changes |= paramsModified
+	}
+	if !util.DeepEqual(ctg.HealthyThresholdCount, dtg.HealthyThresholdCount) {
+		tg.logger.Debugf("HealthyThresholdCount needs to be changed (%v != %v)", log.Prettify(ctg.HealthyThresholdCount), log.Prettify(dtg.HealthyThresholdCount))
+		changes |= paramsModified
+	}
+
+	if !util.DeepEqual(ctg.Matcher, dtg.Matcher) {
+		tg.logger.Debugf("Matcher needs to be changed (%v != %v)", log.Prettify(ctg.Matcher), log.Prettify(ctg.Matcher))
+		changes |= paramsModified
+	}
+
+	if !util.DeepEqual(ctg.UnhealthyThresholdCount, dtg.UnhealthyThresholdCount) {
+		tg.logger.Debugf("UnhealthyThresholdCount needs to be changed (%v != %v)", log.Prettify(ctg.UnhealthyThresholdCount), log.Prettify(dtg.UnhealthyThresholdCount))
+		changes |= paramsModified
+	}
+
+	if *tg.Targets.Current.Hash() != *tg.Targets.Desired.Hash() {
+		tg.logger.Debugf("Targets need to be changed.")
+		changes |= targetsModified
+	}
+
+	if *tg.Tags.Current.Hash() != *tg.Tags.Desired.Hash() {
+		tg.logger.Debugf("Tags need to be changed")
+		changes |= tagsModified
+	}
+
+	if !reflect.DeepEqual(tg.CurrentAttributes.Sorted(), tg.DesiredAttributes.Sorted()) {
+		tg.logger.Debugf("Attributes need to be changed")
+		changes |= attributesModified
+	}
+
+	return changes
 	// These fields require a rebuild and are enforced via TG name hash
 	//	Port *int64 `min:"1" type:"integer"`
 	//	Protocol *string `type:"string" enum:"ProtocolEnum"`
-
-	return false
 }
 
 // Registers Targets (ec2 instances) to TargetGroup, must be called when Current != Desired

--- a/pkg/alb/targetgroups/targetgroups.go
+++ b/pkg/alb/targetgroups/targetgroups.go
@@ -144,7 +144,7 @@ type NewDesiredTargetGroupsOptions struct {
 
 // NewDesiredTargetGroups returns a new targetgroups.TargetGroups based on an extensions.Ingress.
 func NewDesiredTargetGroups(o *NewDesiredTargetGroupsOptions) (TargetGroups, error) {
-	output := o.ExistingTargetGroups
+	var output TargetGroups
 
 	for _, rule := range o.Ingress.Spec.Rules {
 		for _, path := range rule.HTTP.Paths {
@@ -167,8 +167,14 @@ func NewDesiredTargetGroups(o *NewDesiredTargetGroupsOptions) (TargetGroups, err
 				Targets:        o.GetNodes(),
 			})
 
-			// If this target group is already defined, copy the desired state over
-			if i, tg := output.FindById(targetGroup.ID); i >= 0 {
+			// If this target group is already defined, copy the current state to our new TG
+			if i, tg := o.ExistingTargetGroups.FindById(targetGroup.ID); i >= 0 {
+				targetGroup.CurrentAttributes = tg.CurrentAttributes
+				targetGroup.Current = tg.Current
+				targetGroup.Targets.Current = tg.Targets.Current
+				targetGroup.Tags.Current = tg.Tags.Current
+
+				// If there is a current TG ARN we can use it to purge the desired targets of unready instances
 				if tg.Current != nil {
 					targets := []*elbv2.TargetDescription{}
 					for _, instanceID := range targetGroup.Targets.Desired {
@@ -183,10 +189,6 @@ func NewDesiredTargetGroups(o *NewDesiredTargetGroupsOptions) (TargetGroups, err
 					}
 					targetGroup.Targets.Desired = desired
 				}
-				targetGroup.Tags.Desired = tg.Tags.Desired
-				targetGroup.Desired = tg.Desired
-				targetGroup.DesiredAttributes = tg.DesiredAttributes
-				continue
 			}
 
 			output = append(output, targetGroup)

--- a/pkg/alb/targetgroups/targetgroups.go
+++ b/pkg/alb/targetgroups/targetgroups.go
@@ -116,6 +116,13 @@ func NewCurrentTargetGroups(o *NewCurrentTargetGroupsOptions) (TargetGroups, err
 			return nil, err
 		}
 		tg.Targets.Current = current
+
+		v, err := albelbv2.ELBV2svc.DescribeTargetGroupAttributes(&elbv2.DescribeTargetGroupAttributesInput{TargetGroupArn: targetGroup.TargetGroupArn})
+		if err != nil {
+			return nil, err
+		}
+		tg.CurrentAttributes = v.Attributes
+
 		output = append(output, tg)
 	}
 
@@ -157,26 +164,28 @@ func NewDesiredTargetGroups(o *NewDesiredTargetGroupsOptions) (TargetGroups, err
 				Port:           *port,
 				Logger:         o.Logger,
 				SvcName:        path.Backend.ServiceName,
+				Targets:        o.GetNodes(),
 			})
-			targetGroup.Targets.Desired = o.GetNodes()
 
 			// If this target group is already defined, copy the desired state over
 			if i, tg := output.FindById(targetGroup.ID); i >= 0 {
-				targets := []*elbv2.TargetDescription{}
-				for _, instanceID := range targetGroup.Targets.Desired {
-					targets = append(targets, &elbv2.TargetDescription{
-						Id:   instanceID,
-						Port: port,
-					})
+				if tg.Current != nil {
+					targets := []*elbv2.TargetDescription{}
+					for _, instanceID := range targetGroup.Targets.Desired {
+						targets = append(targets, &elbv2.TargetDescription{
+							Id:   instanceID,
+							Port: port,
+						})
+					}
+					desired, err := albelbv2.ELBV2svc.DescribeTargetGroupTargetsForArn(tg.Current.TargetGroupArn, targets)
+					if err != nil {
+						return nil, err
+					}
+					targetGroup.Targets.Desired = desired
 				}
-				desired, err := albelbv2.ELBV2svc.DescribeTargetGroupTargetsForArn(tg.Current.TargetGroupArn, targets)
-				if err != nil {
-					return nil, err
-				}
-
-				tg.Tags.Desired = targetGroup.Tags.Desired
-				tg.Desired = targetGroup.Desired
-				tg.Targets.Desired = desired
+				targetGroup.Tags.Desired = tg.Tags.Desired
+				targetGroup.Desired = tg.Desired
+				targetGroup.DesiredAttributes = tg.DesiredAttributes
 				continue
 			}
 

--- a/pkg/annotations/annotations.go
+++ b/pkg/annotations/annotations.go
@@ -13,6 +13,7 @@ import (
 	"github.com/aws/aws-sdk-go/service/ec2"
 	"github.com/aws/aws-sdk-go/service/elbv2"
 	albec2 "github.com/coreos/alb-ingress-controller/pkg/aws/ec2"
+	albelbv2 "github.com/coreos/alb-ingress-controller/pkg/aws/elbv2"
 	"github.com/coreos/alb-ingress-controller/pkg/config"
 	albprom "github.com/coreos/alb-ingress-controller/pkg/prometheus"
 	"github.com/coreos/alb-ingress-controller/pkg/util/log"
@@ -75,7 +76,7 @@ type Annotations struct {
 	SuccessCodes               *string
 	Tags                       []*elbv2.Tag
 	IgnoreHostHeader           *bool
-	TargetGroupAttributes      []*elbv2.TargetGroupAttribute
+	TargetGroupAttributes      albelbv2.TargetGroupAttributes
 	VPCID                      *string
 	Attributes                 []*elbv2.LoadBalancerAttribute
 }
@@ -649,9 +650,14 @@ func (a *Annotations) setTags(annotations map[string]string) error {
 	return nil
 }
 func (a *Annotations) setTargetGroupAttributes(annotations map[string]string) error {
-	var attrs []*elbv2.TargetGroupAttribute
 	var badAttrs []string
 	rawAttrs := util.NewAWSStringSlice(annotations[targetGroupAttributesKey])
+
+	a.TargetGroupAttributes.Set("deregistration_delay.timeout_seconds", "300")
+	a.TargetGroupAttributes.Set("slow_start.duration_seconds", "0")
+	a.TargetGroupAttributes.Set("stickiness.enabled", "false")
+	a.TargetGroupAttributes.Set("stickiness.lb_cookie.duration_seconds", "86400")
+	a.TargetGroupAttributes.Set("stickiness.type", "lb_cookie")
 
 	for _, rawAttr := range rawAttrs {
 		parts := strings.Split(*rawAttr, "=")
@@ -662,16 +668,13 @@ func (a *Annotations) setTargetGroupAttributes(annotations map[string]string) er
 			badAttrs = append(badAttrs, *rawAttr)
 			continue
 		}
-		attrs = append(attrs, &elbv2.TargetGroupAttribute{
-			Key:   aws.String(parts[0]),
-			Value: aws.String(parts[1]),
-		})
+		a.TargetGroupAttributes.Set(parts[0], parts[1])
 	}
-	a.TargetGroupAttributes = attrs
 
 	if len(badAttrs) > 0 {
 		return fmt.Errorf("Unable to parse `%s` into Key=Value pair(s)", strings.Join(badAttrs, ", "))
 	}
+
 	return nil
 }
 

--- a/pkg/annotations/annotations_test.go
+++ b/pkg/annotations/annotations_test.go
@@ -163,14 +163,16 @@ func TestSetTargetGroupAttributes(t *testing.T) {
 	annotations := &Annotations{}
 	attributes := map[string]string{targetGroupAttributesKey: "deregistration_delay.timeout_seconds=60,stickiness.enabled=true"}
 	err := annotations.setTargetGroupAttributes(attributes)
-	if err != nil || len(annotations.TargetGroupAttributes) != 2 {
+	if err != nil || len(annotations.TargetGroupAttributes) != 5 {
 		t.Errorf("setTargetGroupAttributes - number of attributes incorrect")
 	}
 
-	if err == nil && *annotations.TargetGroupAttributes[0].Key != "deregistration_delay.timeout_seconds" || *annotations.TargetGroupAttributes[0].Value != "60" {
-		t.Errorf("setTargetGroupAttributes - values did not match")
-	}
-	if err == nil && *annotations.TargetGroupAttributes[1].Key != "stickiness.enabled" || *annotations.TargetGroupAttributes[1].Value != "true" {
-		t.Errorf("setTargetGroupAttributes - values did not match")
+	for _, attr := range annotations.TargetGroupAttributes {
+		if *attr.Key == "deregistration_delay.timeout_seconds" && *attr.Value != "60" {
+			t.Errorf("setTargetGroupAttributes - deregistration_delay.timeout_seconds value did not match")
+		}
+		if *attr.Key == "stickiness.enabled" && *attr.Value != "true" {
+			t.Errorf("setTargetGroupAttributes - stickiness.enabled value did not match")
+		}
 	}
 }

--- a/pkg/aws/ec2/ec2.go
+++ b/pkg/aws/ec2/ec2.go
@@ -2,6 +2,7 @@ package ec2
 
 import (
 	"fmt"
+	"os"
 	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
@@ -587,6 +588,10 @@ func (e *EC2) CreateNewInstanceSG(sgName *string, sgID *string, vpcID *string) (
 // the DescribeInstances call to determine its VPC ID.
 func (e *EC2) GetVPCID() (*string, error) {
 	var vpc *string
+
+	if v := os.Getenv("AWS_VPC_ID"); v != "" {
+		return &v, nil
+	}
 
 	// If previously looked up (and not expired) the VpcId will be stored in the cache under the
 	// key 'vpc'.

--- a/pkg/aws/elbv2/elbv2.go
+++ b/pkg/aws/elbv2/elbv2.go
@@ -60,6 +60,18 @@ func (a TargetGroupAttributes) Sorted() TargetGroupAttributes {
 	return a
 }
 
+func (a *TargetGroupAttributes) Set(k, v string) {
+	t := *a
+	for i := range t {
+		if *t[i].Key == k {
+			t[i].Value = aws.String(v)
+			return
+		}
+	}
+
+	*a = append(*a, &elbv2.TargetGroupAttribute{Key: aws.String(k), Value: aws.String(v)})
+}
+
 // ELBV2 is our extension to AWS's elbv2.ELBV2
 type ELBV2 struct {
 	elbv2iface.ELBV2API

--- a/pkg/controller/alb-controller.go
+++ b/pkg/controller/alb-controller.go
@@ -425,6 +425,11 @@ func (ac *albController) GetNodes() util.AWSStringSlice {
 		if _, ok := n.ObjectMeta.Labels["node-role.kubernetes.io/master"]; ok {
 			continue
 		}
+		if s, ok := n.ObjectMeta.Labels["alpha.service-controller.kubernetes.io/exclude-balancer"]; ok {
+			if strings.ToUpper(s) == "TRUE" {
+				continue
+			}
+		}
 		result = append(result, aws.String(n.Spec.ExternalID))
 	}
 	sort.Sort(result)

--- a/pkg/controller/alb-controller.go
+++ b/pkg/controller/alb-controller.go
@@ -22,8 +22,11 @@ import (
 	"k8s.io/ingress/core/pkg/ingress/controller"
 	"k8s.io/ingress/core/pkg/ingress/defaults"
 
+	"strings"
+
 	"github.com/coreos/alb-ingress-controller/pkg/albingress"
 	"github.com/coreos/alb-ingress-controller/pkg/albingresses"
+	"github.com/coreos/alb-ingress-controller/pkg/annotations"
 	"github.com/coreos/alb-ingress-controller/pkg/aws/acm"
 	"github.com/coreos/alb-ingress-controller/pkg/aws/ec2"
 	"github.com/coreos/alb-ingress-controller/pkg/aws/elbv2"
@@ -33,29 +36,28 @@ import (
 	"github.com/coreos/alb-ingress-controller/pkg/config"
 	albprom "github.com/coreos/alb-ingress-controller/pkg/prometheus"
 	"github.com/coreos/alb-ingress-controller/pkg/util/log"
+	albsync "github.com/coreos/alb-ingress-controller/pkg/util/sync"
 	util "github.com/coreos/alb-ingress-controller/pkg/util/types"
 	"github.com/prometheus/client_golang/prometheus"
-	"strings"
-	"github.com/coreos/alb-ingress-controller/pkg/annotations"
 )
 
 // albController is our main controller
 type albController struct {
-	storeLister     ingress.StoreLister
-	recorder        record.EventRecorder
-	ALBIngresses    albingresses.ALBIngresses
-	clusterName     string
-	albNamePrefix   string
-	IngressClass    string
-	lastUpdate      time.Time
-	albSyncInterval time.Duration
-	mutex           sync.RWMutex
-	awsChecks       map[string]func() error
-	poller	        func (*albController)
-	initialSync     func (*albController)
-	syncer	        func (*albController)
-	classNameGetter func (*controller.GenericController) string
-	recorderGetter  func (*controller.GenericController) record.EventRecorder
+	storeLister       ingress.StoreLister
+	recorder          record.EventRecorder
+	ALBIngresses      albingresses.ALBIngresses
+	clusterName       string
+	albNamePrefix     string
+	IngressClass      string
+	lastUpdate        time.Time
+	albSyncInterval   time.Duration
+	mutex             albsync.RWMutex
+	awsChecks         map[string]func() error
+	poller            func(*albController)
+	initialSync       func(*albController)
+	syncer            func(*albController)
+	classNameGetter   func(*controller.GenericController) string
+	recorderGetter    func(*controller.GenericController) record.EventRecorder
 	annotationFactory annotations.AnnotationFactory
 }
 

--- a/pkg/util/log/log.go
+++ b/pkg/util/log/log.go
@@ -44,7 +44,12 @@ func New(name string) *Logger {
 
 // Debugf will print debug messages if debug logging is enabled
 func (l *Logger) Debugf(format string, args ...interface{}) {
-	debugf(format, l.name, args...)
+	debugf(format, l.name, 2, args...)
+}
+
+// DebugLevelf will print debug messages if debug logging is enabled
+func (l *Logger) DebugLevelf(level int, format string, args ...interface{}) {
+	debugf(format, l.name, level, args...)
 }
 
 // Infof will print info level messages
@@ -73,12 +78,12 @@ func (l *Logger) Exitf(format string, args ...interface{}) {
 }
 
 // debugf will print debug messages if debug logging is enabled
-func debugf(format, ingressName string, args ...interface{}) {
+func debugf(format, ingressName string, level int, args ...interface{}) {
 	if logLevel > INFO {
 		ingressName = leftBracket + ingressName + rightBracket
 		prefix := fmt.Sprintf("%s %s %s: ", identifier, ingressName, debugLevel)
 		for _, line := range strings.Split(fmt.Sprintf(format, args...), "\n") {
-			glog.InfoDepth(2, prefix, line)
+			glog.InfoDepth(level, prefix, line)
 		}
 	}
 }

--- a/pkg/util/sync/sync.go
+++ b/pkg/util/sync/sync.go
@@ -1,0 +1,39 @@
+package sync
+
+import (
+	"sync"
+
+	"github.com/coreos/alb-ingress-controller/pkg/util/log"
+)
+
+var logger *log.Logger
+
+func init() {
+	logger = log.New("sync")
+}
+
+type RWMutex struct {
+	m sync.RWMutex
+}
+
+func (r *RWMutex) RLock() {
+	logger.DebugLevelf(3, "Requesting RLock.")
+	r.m.RLock()
+	logger.DebugLevelf(3, "RLock was available.")
+}
+
+func (r *RWMutex) Lock() {
+	logger.DebugLevelf(3, "Requesting Lock.")
+	r.m.Lock()
+	logger.DebugLevelf(3, "Lock was available.")
+}
+
+func (r *RWMutex) RUnlock() {
+	r.m.RUnlock()
+	logger.DebugLevelf(3, "RUnlock")
+}
+
+func (r *RWMutex) Unlock() {
+	r.m.Unlock()
+	logger.DebugLevelf(3, "Unlock")
+}

